### ISSUE TITLE
add_today.rb filter hangs running via pandocomatic

### ISSUE THREAD

**Issue by iandol** [htdebeer/paru#10]
Hi, using the new metadata.yaml version of the add_today.rb, my pandocomatic compiles are hanging, and I have to do a CTRL+C to force break:

```
👉  pandocomatic --debug "SNN-Attention-V1.4.md"
^Cpandoc: Error running filter /Users/ian/.pandoc/filters/add_today.rb
user interrupt
/Users/ian/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/paru-0.2.4.8/lib/paru/pandoc.rb:161:in `read': Interrupt
	from /Users/ian/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/paru-0.2.4.8/lib/paru/pandoc.rb:161:in `block in convert'
	from /Users/ian/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/paru-0.2.4.8/lib/paru/pandoc.rb:158:in `popen'
	from /Users/ian/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/paru-0.2.4.8/lib/paru/pandoc.rb:158:in `convert'
	from /Users/ian/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/pandocomatic-0.1.4.1/lib/pandocomatic/command/convert_file_command.rb:157:in `pandoc'
	from /Users/ian/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/pandocomatic-0.1.4.1/lib/pandocomatic/command/convert_file_command.rb:92:in `convert_file'
	from /Users/ian/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/pandocomatic-0.1.4.1/lib/pandocomatic/command/convert_file_command.rb:59:in `run'
	from /Users/ian/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/pandocomatic-0.1.4.1/lib/pandocomatic/command/command.rb:87:in `execute'
	from /Users/ian/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/pandocomatic-0.1.4.1/lib/pandocomatic/command/convert_file_multiple_command.rb:81:in `block in execute'
	from /Users/ian/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/pandocomatic-0.1.4.1/lib/pandocomatic/command/convert_file_multiple_command.rb:80:in `each'
	from /Users/ian/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/pandocomatic-0.1.4.1/lib/pandocomatic/command/convert_file_multiple_command.rb:80:in `execute'
	from /Users/ian/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/pandocomatic-0.1.4.1/lib/pandocomatic/pandocomatic.rb:108:in `run'
	from /Users/ian/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/pandocomatic-0.1.4.1/bin/pandocomatic:3:in `<top (required)>'
	from /Users/ian/.rbenv/versions/2.4.1/bin/pandocomatic:22:in `load'
	from /Users/ian/.rbenv/versions/2.4.1/bin/pandocomatic:22:in `<main>'
```

other paru filters are working.

EDIT: this filter also contains a `<<~` HEREDOC, but making it `<<-` and it still hangs, and this was tested on Ruby 2.4.1 (installed via rbenv)

```ruby
#!/usr/bin/env ruby
## Add today's date to the metadata
require "paru/filter"
require "date"

Paru::Filter.run do 
    metadata.yaml <<-YAML
---
date: #{Date.today.to_s}
...
    YAML
end
```

**Comment by iandol** [comment]
Actually if I leave it for around 3 minutes, it does finally finish, so this is a performance issue. I also tested with my Ruby V2.0 system install with the same results.

**Comment by htdebeer** [comment]
This is an odd one! If I run

    time pandoc --from markdown --to markdown --standalone --filter ~/Projects/paru/examples/filters/add_today.rb test2.md

it gives back 

    real    0m0,660s
    user    0m0,092s
    sys     0m0,032s

Then I tried it to run the filter via pandocomatic like you did, and it gives a similar result.

So, I am wondering, what is the difference between your setup and mine?

- what version of pandoc are you using? (I'm using 1.19.2.1)
- what operating system are you using? (I'm using ruby 2.3.3p222 (2016-11-21) [x86_64-linux-gnu])
- what happens if you run the filter directly with pandoc rather than through pandocomatic?

A potential issue with setting the metadata might be that it is running pandoc to strip the metadata. Maybe there is some issue with running pandoc while running pandoc?

**Comment by htdebeer** [comment]
A potential problem might be that the filter is run for each node in pandoc's AST tree of your document. Is your document quite large? Or rather, what happens when you run the filter on a file with almost nothing in it?

**Comment by iandol** [comment]
The document that hangs is quite large (~20KB), but even with a minimal example[^1] I see a substantial delay (~1.3s vs. 0.02s):

```shell
👉  time pandoc -F add_today.rb -t markdown min.md
pandoc -F add_today.rb -t markdown min.md  1.32s user 1.14s system 81% cpu 3.014 total

👉  time pandoc -t markdown min.md
pandoc -t markdown min.md  0.02s user 0.01s system 71% cpu 0.042 total

👉  ruby -v
ruby 2.4.1p111 (2017-03-22 revision 58053) [x86_64-darwin16]

👉  pandoc -v | head -n 3
pandoc 1.19.2.1
Compiled with pandoc-types 1.17.0.5, texmath 0.9, skylighting 0.1.1.4
Default user data directory: /Users/ian/.pandoc

pandocomatic (0.1.4.1)
paru (0.2.4.8, 0.2.4.7)
```

I will try to use byebug tomorrow to trace if where most time is taken, not sure if there are other ways to generate traces from Ruby?

----
[^1]: 
```markdown
---
title: "Minimalist Example"
author: John Doe
pandocomatic_:
  use-template: eisvogel
---

# Introduction #
Lørem ipsum dolør sit amet[@shipp2013], eu ipsum movet vix[^fn1], veniam låoreet posidonium te eøs, eæm in veri eirmod. Sed illum ⬄ minimum at, est mægna alienum mentitum ne. Amet equidem Lgn sit ex. Ludus øfficiis suåvitate sea in, ius utinam vivendum no, mei nostrud necessitatibus te[@clark2013]?

# References #


[^fn1]: test footnote
```

If I comment out the code within the run block it is fast again:

```ruby
#!/usr/bin/env ruby
require "paru/filter"
require "date"

Paru::Filter.run do 
#metadata.yaml <<-YAML
#---
#date: #{Date.today.to_s}
#...
#YAML
end
```
```
👉  time pandoc -F add_today.rb -t markdown min.md
pandoc -F add_today.rb -t markdown min.md  0.14s user 0.04s system 81% cpu 0.227 total
```

**Comment by htdebeer** [comment]
I'll look into a way to monitor the performance / trace in a filter today. I'll also try to install rbenv and ruby 2.4 and see if that makes a difference on my machine. 

I do wonder, however, other filters are fine? Only changing the metadata grinds everything to a halt?

**Comment by iandol** [comment]
Yes, your removeHR filter works fine, it is only the metadata filter.

```ruby
#!/usr/bin/env ruby
DEBUG = true
if DEBUG
	PORT = 8991 
	require 'byebug/core'
	require 'byebug'
	STDOUT.puts "\n!!!---DEBUG Server started on localhost:#{PORT} @ " + Time.now.to_s
	Byebug.wait_connection = true
	Byebug.start_server('localhost', PORT)
	byebug
end

require "paru/filter"
require "date"

Paru::Filter.run do 
metadata.yaml <<-YAML
---
date: #{Date.today.to_s}
...
YAML
end
```

This is a modified filter that uses a remote byebug setup. You run Pandoc with the filter and in a new terminal you use `byebug -R localhost:8991` to put you into a debug session. `s` steps into methods and you can follow the route through. I couldn't easily spot anything when I tried it but the trace goes deep...

**Comment by htdebeer** [comment]
Yes. Although I am not sure this is causing the issue, but now that I've just installed rbenv and Ruby 2.4.1, it is complaining about circular references everywhere. I see if I can solve those first.

**Comment by htdebeer** [comment]
This seems to be the result of a new version of rake I installed together with rbenv. I can disable the warnings for now.

**Comment by iandol** [comment]
https://github.com/ruby-prof/ruby-prof allows you to set it just around the filter then save the results to file...

**Comment by iandol** [comment]
```ruby
#!/usr/bin/env ruby

require "paru/filter"
require "date"
require 'ruby-prof'

RubyProf.start
Paru::Filter.run do 
metadata.yaml <<-YAML
---
date: #{Date.today.to_s}
...
YAML
end
result = RubyProf.stop

printer = RubyProf::MultiPrinter.new(result)
printer.print(:path => ".", :profile => "addToday")
```

**Comment by htdebeer** [comment]
Thanks. This is useful. I have been able to reproduce it with a new test file I generated, which takes about 11 seconds to just add the metadata. The problem is that the metadata is being set for each AST node the filter encounters. Which is a lot!. Particularly because I am using pandoc twice to convert the YAML to a pandoc AST each time.

But that is also something that I should be able to make more efficient.

**Comment by iandol** [comment]
![screen shot 2017-06-04 at 10 06 53](https://cloud.githubusercontent.com/assets/153905/26758332/b8640ff8-490d-11e7-9274-aa51b4a5e75f.png)

As you say, this is where almost all the time is spent (155secs out of 168secs).  L237 of filter.rb calling L158 of Pandoc.rb which is an IO.popen. Why does it need to set metadata of each node?

**Comment by htdebeer** [comment]
It does not have to do so. Currently I have mixed in the metadata with the document, but it makes sense to treat them separately. I am looking into the following avenues to optimize the metadata behavior:

- preconvert the YAML you want to set in the metadata outside the filter and then set that. For example
      
      data = Paru::Metadata.yaml "my yaml string"
      Paru::Filter.run do
          metadata.set("my.property.key", data) 
      end

  this would still be run for each node, but the pandoc conversions will only be run once  
- remember if the metadata has already been set and only set it just the one time or when the metadata you want to set is different.
- add "prepare" and "teardown" sections to a filter in which you can set metadata as well and you have access to the AST

**Comment by htdebeer** [comment]
I have implemented caching for metadata. This should prevent the issue. I have added it to paru-0.2.5d, which can be found in the releases directory. Further improvements to setting the metadata will follow in the upcoming 0.2.5 release.

**Comment by iandol** [comment]
OK, looks fixed, comparing (1) no filter (**0.13**s), (2) empty paru filter (**0.3**s), (3) add date filter(**0.6**s):

```bash
👉  time pandoc -s -t latex -o blank.tex blank.md
pandoc -s -t latex -o blank.tex blank.md  0.11s user 0.02s system 91% cpu 0.133 total

👉  time pandoc -s -t latex -F noop -o blank.tex blank.md
pandoc -s -t latex -F noop -o blank.tex blank.md  0.25s user 0.04s system 92% cpu 0.310 total

👉  time pandoc -s -t latex -F addToday -o blank.tex blank.md
pandoc -s -t latex -F addToday -o blank.tex blank.md  0.27s user 0.06s system 51% cpu 0.632 total
```

I think this issue is fixed.
